core/src/main/java/org/ehcache/events/CacheEventDispatcherImpl.java

```diff
@@ -95,157 +125,107 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
    */
   @Override
   public void deregisterCacheEventListener(CacheEventListener<? super K, ? super V> listener) {
-    if (!registeredListeners.remove(new EventListenerWrapper(listener,
-        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class)))) {
-      throw new IllegalStateException("Unknown cache event listener: " + listener);
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener);
+
+    if (!removeWrapperFromSet(wrapper, aSyncListenersSet)) {
+      if (!removeWrapperFromSet(wrapper, syncListenersSet)) {
+        throw new IllegalStateException("Unknown cache event listener: " + listener);
+      }
     }
-    if (!hasListeners()) {
-      store.disableStoreEventNotifications();
+  }
+
+  private boolean removeWrapperFromSet(EventListenerWrapper wrapper, Set<EventListenerWrapper> listenersSet) {
+    for (EventListenerWrapper listenerWrapper : listenersSet) {
```

Comment:
The wrapper I have to remove only contains the _public_ listener, not its configurations (ordered or not, async / sync), so I need to access the one in the set to do proper accounting.